crates/ruff_linter/src/rules/ruff/rules/control_var_used_after_block.rs

```diff
@@ -0,0 +1,143 @@
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for variables defined in `for` loops that are used outside of their
+/// respective blocks.
+///
+/// ## Why is this bad?
+/// Usage of of a control variable outside of the block they're defined in will probably
+/// lead to flawed logic in a way that will likely cause bugs. The variable might not
+/// contain what you expect.
+///
+/// In Python, unlike many other languages, `for` loops don't define their own scopes.
+/// Therefore, usage of the control variables outside of the block will be the the value
+/// from the last iteration until re-assigned.
+///
+/// While this mistake is easy to spot in small examples, it can be hidden in larger
+/// blocks of code, where the the loop and downstream usage may not be visible at the
+/// same time.
+///
+/// ## Example
+/// ```python
+/// for x in range(10):
+///     pass
+///
+/// print(x)  # prints 9
+/// ```
+#[violation]
+pub struct ControlVarUsedAfterBlock {
+    control_var_name: String,
+    block_kind: BlockKind,
+}
+
+impl Violation for ControlVarUsedAfterBlock {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let ControlVarUsedAfterBlock {
+            control_var_name,
+            block_kind,
+        } = self;
+
+        format!("{block_kind} variable {control_var_name} is used outside of block")
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum BlockKind {
+    For,
+}
+
+impl fmt::Display for BlockKind {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            BlockKind::For => fmt.write_str("`for` loop"),
+        }
+    }
+}
+
+/// Based on wemake-python-styleguide (WPS441) to forbid control variables after the block body.
+pub(crate) fn control_var_used_after_block(
+    checker: &Checker,
+    scope: &Scope,
+    diagnostics: &mut Vec<Diagnostic>,
+) {
+    // Keep track of node_ids we know are used in the same block. This helps us when
+    // people use the same variable name in multiple blocks.
+    let mut known_good_reference_node_ids: Vec<NodeId> = Vec::new();
+
+    let all_bindings: Vec<(&str, BindingId)> = scope.all_bindings().collect();
+    // We want to reverse the bindings so that we iterate in source order and shadowed
+    // bindings come first.
+    let reversed_bindings = all_bindings.iter().rev();
+
+    // Find for-loop variable bindings
+    for (&name, binding) in reversed_bindings
+        .map(|(name, binding_id)| (name, checker.semantic().binding(*binding_id)))
+        .filter_map(|(name, binding)| {
+            if binding.kind.is_loop_var() {
+                return Some((name, binding));
+            }
+
+            None
+        })
+    {
+        let binding_statement = binding.statement(checker.semantic()).unwrap();
+        let binding_source_node_id = binding.source.unwrap();
+        // The node_id of the for-loop that contains the binding
```

Comment:
I feel like I might need some help on these.

### 1. `binding.statement(checker.semantic()).unwrap();`

What scenarios would a variable `binding.statement` be `None`? Feel like this might be a situation where it's a problem with `ruff` itself and we should give up. How do I represent that?


### 2. `binding.source.unwrap();`

I think we're safe here as how can you bind a variable without the `source`? This again feels it would be a problem with `ruff` itself and we should give up. How do I represent that?

Looking wherever we create a new `Binding { ... }`, it seems like [`source` is `None`  with `builtin`'s](https://github.com/astral-sh/ruff/blob/2e2b1b460f4de25e630cde69b4b49577bc5134b4/crates/ruff_python_semantic/src/model.rs#L229-L241).

I don't think it's possible for us to deal with variable bindings for a `builtin`? (please double-check with your own knowledge)

Although I guess the following example works in the Python interpreter but it doesn't seem to pop up as a variable `binding` at all in `ruff`:

```python
import builtins

for builtins.x in [1,2,3]:
    _ = x
    pass

# 3
print(x)
```

### 3. `reference.expression_id().unwrap()`

This one says:

[`reference.rs#L16-L18`](https://github.com/astral-sh/ruff/blob/2e2b1b460f4de25e630cde69b4b49577bc5134b4/crates/ruff_python_semantic/src/reference.rs#L16-L18)
```
/// The expression that the reference occurs in. `None` if the reference is a global
/// reference or a reference via an augmented assignment.
```

I'm not sure what it means by "global reference" as it seems to work fine with our `global_var` example in the fixture file.

I don't think we can run into an [augmented assignment](https://docs.python.org/3/reference/simple_stmts.html#augmented-assignment-statements) (`x += 1`) in the for-loop binding syntax? (please double-check with your own knowledge)

It seems like this can only be `None` if the `SemanticModel.node_id` is [`None` which happens on on `new(...)`](https://github.com/astral-sh/ruff/blob/2e2b1b460f4de25e630cde69b4b49577bc5134b4/crates/ruff_python_semantic/src/model.rs#L150-L155) and gets filled in as soon as we `push_node` something onto the stack. If we're iterating over any source code, it feels like something would be pushed onto the stack.

Reply:
### 1. `binding.statement(checker.semantic()).unwrap();`

Answered in a discussion below:

> The `statement` can be `None` if the binding comes from a built in. Let's add a comment explaining that calling `unwrap` is ok, because the rule only looks at for loop bindings.
>
> *-- https://github.com/astral-sh/ruff/pull/11769#discussion_r1708959174*

Thankfully, we can avoid this `unwrap()` altogether if we grab the `binding_statement` a different way (see https://github.com/astral-sh/ruff/pull/11769#discussion_r1708963581)


### `2. binding.source.unwrap();`

(no discussion yet)

### 3. `reference.expression_id().unwrap()`

Answered in a discussion below which seems in line with what I was finding.

> According to the documentation, the expression can be `None` when
> 
> ```
> /// The expression that the reference occurs in. `None` if the reference is a global
> /// reference or a reference via an augmented assignment.
> ```
> 
> As far as I can tell, this doesn't apply here. It would be good to add a comment here explaining why calling `unwrap` is safe.
> 
> *-- https://github.com/astral-sh/ruff/pull/11769#discussion_r1708953364*